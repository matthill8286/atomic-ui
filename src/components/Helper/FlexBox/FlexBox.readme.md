## FlexBox

A simple implementation of the flex-box api.

In this example by default the debug is enabled.

Basic Example with default values:
```html
<FlexBox>
    <FlexItem>First Item</FlexItem>
    <FlexItem>Second Item</FlexItem>
</FlexBox>
```

&nbsp;

----------

&nbsp;

#### FlexBox API

Container element props:

| Props         |      values                 |  default |
|---------------|:---------------------------:|---------:|
|    debug      |  true\|false              |  false   |
| display       |    'flex' \| 'inline-flex'  |   'flex' |
| flexDirection | 'row' \| 'row-reverse' \| 'column' \| 'column-reverse' |  'row' |
| flexWrap      | 'nowrap' \| 'wrap' \| 'wrap-reverse' | 'wrap' |
| justifyContent | 'flex-start' \| 'flex-end' \| 'center' \| 'space-between' \| 'space-around' | 'flex-start' |
| alignContent  | 'flex-start' \| 'flex-end' \| 'center' \| 'baseline' \| 'stretch' | 'flex-start' |
| alignItems    | 'flex-start' \| 'flex-end' \| 'center' \| 'space-between' \| 'space-around' \| 'stretch' | 'flex-start'

&nbsp;

----------

&nbsp;

#### FlexItem API

Props:

| Props   |      values      |  default |
|----------|:-------------:|------:|
| debug   |  true\|false   | false |
| order |    number   |   0 |
| flex | string |    '0 1 auto' |
| alignSelf | 'auto' \| 'flex-start' \| 'flex-end' \| 'center' \| 'baseline' \| 'stretch' | 'auto' |

&nbsp;

<div>
For more advanced scenarios play with this <button onclick="document.getElementById('iframe-flexboxes').style.display = 'block'">example</button>
</div>

&nbsp;

<div id="iframe-flexboxes" style="display: none;">
  <iframe src="https://the-echoplex.net/flexyboxes/" frameborder="0" allowfullscreen="allowfullscreen" width="100%" height="800"></iframe>
</div>
