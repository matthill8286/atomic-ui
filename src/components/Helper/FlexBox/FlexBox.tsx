import React, { FC } from 'react'
import { media } from '@/styles/media'
import { css, styled } from '@/styles/styled'

export type FlexAlignItemsOptions =
  | 'flex-start'
  | 'flex-end'
  | 'center'
  | 'space-between'
  | 'space-around'
  | 'stretch'

export interface FlexBoxProps {
  debug?: boolean
  display?: 'flex' | 'inline-flex'
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse'
  flexWrap?: 'nowrap' | 'wrap' | 'wrap-reverse'
  justifyContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around'
  alignContent?: 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
  alignItems?: FlexAlignItemsOptions
}

const StyledBox = styled.div<FlexBoxProps>`
  ${({ debug }) =>
    debug &&
    css`
      border: 1px solid green;
      padding: 8px;
    `}
  display: ${({ display }) => display};
  flex-direction: ${({ flexDirection }) => flexDirection};
  ${({ flexWrap }) => flexWrap && `flex-wrap: ${flexWrap}`};
  ${({ justifyContent }) => justifyContent && `justify-content: ${justifyContent}`};
  ${({ alignContent }) => alignContent && `align-content: ${alignContent}`};
  ${({ alignItems }) => alignItems && `align-items: ${alignItems}`};
`

export const FlexBox: FC<FlexBoxProps> = props => {
  const {
    debug = false,
    display = 'flex',
    flexDirection = 'row',
    flexWrap,
    justifyContent,
    alignContent,
    alignItems,
    children,
    ...otherProps
  } = props

  return (
    <StyledBox
      debug={debug}
      display={display}
      flexDirection={flexDirection}
      flexWrap={flexWrap}
      justifyContent={justifyContent}
      alignContent={alignContent}
      alignItems={alignItems}
      {...otherProps}>
      {children}
    </StyledBox>
  )
}

FlexBox.displayName = 'FlexBox'

export interface FlexItemProps {
  debug?: boolean
  order?: number
  flex?: string
  alignSelf?: 'auto' | 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch'
  padding?: boolean
}

const StyledFlexItem = styled.div<FlexItemProps>`
  ${({ debug }) =>
    debug &&
    css`
      border: 1px solid blue;
      padding: 8px;
    `}
  order: ${({ order }) => order};
  flex: ${({ flex }) => flex};
  align-self: ${({ alignSelf }) => alignSelf};
  padding: ${({ theme, padding }) => (padding ? `0 ${theme.spacing.base.xs}` : 0)};
`

export const FlexItem: FC<FlexItemProps> = props => {
  const { children, debug, order = 0, flex = '0 1 auto', alignSelf = 'auto', ...otherProps } = props

  return (
    <StyledFlexItem debug={debug} order={order} flex={flex} alignSelf={alignSelf} {...otherProps}>
      {children}
    </StyledFlexItem>
  )
}

FlexItem.displayName = 'FlexItem'
