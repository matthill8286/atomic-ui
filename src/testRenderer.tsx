import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
