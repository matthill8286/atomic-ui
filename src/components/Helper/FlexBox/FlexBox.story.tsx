import { boolean, number, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { FlexBox, FlexBoxProps, FlexItem, FlexItemProps } from './FlexBox'
import readme from './FlexBox.readme.md'

storiesOf('Design System/Helper/FlexBox', module).add(
  'Default',
  () => {
    const flexBoxKnobs: FlexBoxProps = {
      debug: boolean('box debug', true),
      display: select('box display', ['flex', 'inline-flex'], 'flex'),
      flexDirection: select(
        'box flexDirection',
        ['row', 'row-reverse', 'column', 'column-reverse'],
        'row'
      ),
      flexWrap: select('box:flexWrap', ['nowrap', 'wrap', 'wrap-reverse'], 'wrap'),
      justifyContent: select(
        'box justifyContent',
        ['flex-start', 'flex-end', 'center', 'space-between', 'space-around'],
        'flex-start'
      ),
      alignContent: select(
        'box alignContent',
        ['flex-start', 'flex-end', 'center', 'baseline', 'stretch'],
        'flex-start'
      ),
      alignItems: select(
        'box alignItems',
        ['flex-start', 'flex-end', 'center', 'space-between', 'space-around', 'stretch'],
        'flex-start'
      ),
    }
    const firstFlexItemProps: FlexItemProps = {
      debug: boolean('first_item debug', true),
      order: number('first_item order', 0),
      flex: text('first_item flex', '0 1 auto'),
      alignSelf: select(
        'first_item alignSelf',
        ['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch'],
        'auto'
      ),
    }
    const secondFlexItemProps: FlexItemProps = {
      debug: boolean('second_item debug', true),
      order: number('second_item order', 0),
      flex: text('second_item flex', '0 1 auto'),
      alignSelf: select(
        'second_item alignSelf',
        ['auto', 'flex-start', 'flex-end', 'center', 'baseline', 'stretch'],
        'auto'
      ),
    }
    return (
      <FlexBox {...flexBoxKnobs}>
        <FlexItem {...firstFlexItemProps}>First Item</FlexItem>
        <FlexItem {...secondFlexItemProps}>Second Item</FlexItem>
      </FlexBox>
    )
  },
  { info: readme }
)
