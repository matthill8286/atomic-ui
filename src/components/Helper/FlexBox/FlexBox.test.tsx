import * as React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { FlexBox, FlexBoxProps, FlexItem, FlexItemProps } from './FlexBox'

describe('FlexBox', () => {
  const props: FlexBoxProps = {}

  it('renders correctly', () => {
    const tree = renderWithTheme(<FlexBox {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders with direction column', () => {
    const tree = renderWithTheme(<FlexBox {...props} flexDirection="column" />)
    expect(tree).toMatchSnapshot()
  })

  it('renders with flex wrap', () => {
    const tree = renderWithTheme(<FlexBox {...props} flexWrap="wrap" />)
    expect(tree).toMatchSnapshot()
  })
})

describe('FlexItem', () => {
  const props: FlexItemProps = {}

  it('renders correctly', () => {
    const tree = renderWithTheme(<FlexItem {...props} />)
    expect(tree).toMatchSnapshot()
  })
})
