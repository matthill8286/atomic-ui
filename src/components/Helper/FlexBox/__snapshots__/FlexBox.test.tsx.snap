// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexBox renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<div
  className="c0"
  display="flex"
/>
`;

exports[`FlexBox renders with direction column 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
  display="flex"
/>
`;

exports[`FlexBox renders with flex wrap 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

<div
  className="c0"
  display="flex"
/>
`;

exports[`FlexItem renders correctly 1`] = `
.c0 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  padding: 0;
}

<div
  className="c0"
  order={0}
/>
`;
